### PR TITLE
irmin-pack: refactor and reduce duplication in common signatures

### PR DESCRIPTION
**Pull request by CraigFe**

### CODE REVIEW

File: src/irmin-pack/s.ml [hunk 1]
```diff
@@ -14,95 +14,49 @@
...
+  type key
 
-  val next_upper : 'a t -> [ `Read ] U.t
+  val integrity_check :
+    offset:int64 -> length:int -> key -> _ t -> (unit, integrity_error) result
+end
 
-  val lower : 'a t -> [ `Read ] L.t
+module type LAYERED_GENERAL = sig
+  type 'cap t
```
Comment (icristescu) [inline src/irmin-pack/s.ml:35]: I prefer `a t` here instead of `cap`, if it's ok with you.

File: src/irmin-pack/pack_intf.ml [hunk 2]
```diff
@@ -82,10 +83,9 @@ module type S = sig
...
       are not removed. *)
 
-  type integrity_error = [ `Wrong_hash | `Absent_value ]
+  include Sigs.CHECKABLE with type 'a t := 'a t and type key := key
 
-  val integrity_check :
-    offset:int64 -> length:int -> key -> 'a t -> (unit, integrity_error) result
+  include Sigs.CLOSEABLE with type 'a t := 'a t
 
   val close : 'a t -> unit Lwt.t
```
Comment (icristescu) [inline src/irmin-pack/pack_intf.ml:90]: We can remove this `close`

**Review by icristescu — APPROVED** [review]
Thanks, this is neat! I have just a few remarks.

File: src/irmin-pack/s.ml (hunk 1 above)
Comment (CraigFe) [reply src/irmin-pack/s.ml:35]: Sure :+1:

File: src/irmin-pack/pack_intf.ml (hunk 2 above)
Comment (CraigFe) [reply src/irmin-pack/pack_intf.ml:90]: Indeed, thanks.

File: src/irmin-pack/inode_layers_intf.ml [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    offset:int64 ->
+    length:int ->
+    layer:layer_id ->
+    key ->
+    'a t ->
+    (unit, S.integrity_error) result
+end
+
+module type Inode_layers = sig
+  type nonrec layer_id = layer_id
```
[the remark this answers is not in the document]
Comment (CraigFe) [reply src/irmin-pack/inode_layers_intf.ml:59]: If we don't add `layer_id` to `inode_layers.mli` by putting it here, other modules would have to refer to this type as `Inode_layers_intf.layer_id`. As it happens, other modules are not referring to this type at all and are simply duplicating its definition; that looks like an opportunity for some more de-duplication.

[the remark this answers is not in the document]
Comment (CraigFe) [reply src/irmin-pack/inode_layers_intf.ml]: Done in https://github.com/mirage/irmin/pull/1107/commits/8f2165b6ce35924676229e28281fdaf6cdd55f55, thanks!

**Review by pascutto — APPROVED** [review]

**Pull request merged**